control/velocity_controller_lqr/scripts/velocity_controller_lqr_node.py

```diff
@@ -0,0 +1,189 @@
+
+        i_surge = self.get_parameter("i_surge").value
+        i_pitch = self.get_parameter("i_pitch").value
+        i_yaw = self.get_parameter("i_yaw").value
+
+        i_weight = self.get_parameter("i_weight").value
+        max_force = self.get_parameter("max_force").value
+
+        inertia_matrix_flat = self.get_parameter("inertia_matrix").value
+        inertia_matrix = np.array(inertia_matrix_flat).reshape((3, 3))
+
+        parameters = np.array(
+            [
+                q_surge,
+                q_pitch,
+                q_yaw,
+                r_surge,
+                r_pitch,
+                r_yaw,
+                i_surge,
+                i_pitch,
+                i_yaw,
+                i_weight,
+                max_force,
+            ]
+        )
+
+        for i in range(len(parameters - 2)):
+            assert parameters[i] > 0, f"Parameter {i} is negative"
+        assert 0 <= parameters[-1] <= 99.9, "Max force must be in the range [0, 99.9]."
+
+        return parameters, inertia_matrix
+
+    # ---------------------------------------------------------------Callback Functions---------------------------------------------------------------
+
+    def nucleus_callback(
+        self, msg: Odometry
+    ):  # callback function to read data from nucleus
+        dummy, self.states[1], self.states[2] = LQRController.quaternion_to_euler_angle(
+            msg.pose.pose.orientation.w,
+            msg.pose.pose.orientation.x,
+            msg.pose.pose.orientation.y,
+            msg.pose.pose.orientation.z,
+        )
+
+        self.states[0] = msg.twist.twist.linear.x
+
+        self.coriolis_matrix = self.controller.calculate_coriolis_matrix(
+            msg.twist.twist.angular.y,
+            msg.twist.twist.angular.z,
+            msg.twist.twist.linear.y,
+            msg.twist.twist.linear.z,
+        )  # coriolis matrix
+
+    def guidance_callback(
+        self, msg: LOSGuidance
+    ):  # Function to read data from guidance
+        self.guidance_values[0] = msg.surge
+        self.guidance_values[1] = msg.pitch
+        self.guidance_values[2] = msg.yaw
+
+    # ---------------------------------------------------------------Publisher Functions---------------------------------------------------------------
+
+    def control_loop(self):  # The LQR controller function
+        msg = Wrench()
+
+        u = self.controller.calculate_lqr_u(
+            self.coriolis_matrix, self.states, self.guidance_values
+        )
+        msg.force.x = float(u[0])
+        msg.torque.y = float(u[1])
+        msg.torque.z = float(u[2])
+
+        # Publish the control input
+        self.publisherLQR.publish(msg)
+
+
+# ---------------------------------------------------------------Main Function---------------------------------------------------------------
+
+
+def main(args=None):  # main function
+    rclpy.init(args=args)
+    node = LinearQuadraticRegulator()
+    rclpy.spin(node)
```

Comment:
Consider using [MultiThreadedExecutor]https://docs.ros.org/en/humble/Concepts/Intermediate/About-Executors.htmll)

Reply:
why htmll is it like new html